Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1247,6 +1247,112 @@ The following approach can be used in order to patch the JavaScript file:
 
 #### Dynamic Instrumentation
 
+##### Information Gathering
+
+In this section we will learn about how to use Frida to obtain information about a running application.
+
+###### Getting Loaded Classes and their Methods
+
+In the Frida REPL to access Java runtime, `Java` command can be used to access information within the running app. Unlike `ObjC` command for iOS, for Java, the code need to be called via `Java.perform()` function. Thus, the best way to get list of loaded Java classes and their corresponding methods and field is to use Frida scripts. One such script is listed below.
+
+```
+// Get list of loaded Java classes and methods
+
+Java.perform(function() {
+    Java.enumerateLoadedClasses({
+        onMatch: function(className) {
+            console.log(className);
+            describeJavaClass(className);
+        },
+        onComplete: function() {}
+    });
+});
+
+// Get the methods and fields
+function describeJavaClass(className) {
+  var jClass = Java.use(className);
+  console.log(JSON.stringify({
+    _name: className,
+    _methods: Object.getOwnPropertyNames(jClass.__proto__).filter(function(m) {
+      return !m.startsWith('$') // filter out Frida related special properties
+        || m == 'class' || m == 'constructor' // optional
+    }),
+    _fields: jClass.class.getFields().map(function(f) {
+      return( f.toString());
+    })
+  }, null, 2));
+}
+
+
+// Output
+
+[Huawei Nexus 6P::sg.vantagepoint.helloworldjni]->
+...
+
+com.scottyab.rootbeer.sample.MainActivity
+{
+  "_name": "com.scottyab.rootbeer.sample.MainActivity",
+  "_methods": [
+  ...
+    "beerView",
+    "checkRootImageViewList",
+    "floatingActionButton",
+    "infoDialog",
+    "isRootedText",
+    "isRootedTextDisclaimer",
+    "mActivity",
+    "GITHUB_LINK"
+  ],
+  "_fields": [
+    "public static final int android.app.Activity.DEFAULT_KEYS_DIALER",
+...
+
+com.scottyab.rootbeer.sample.CheckRootTask$OnCheckRootFinishedListener
+{
+  "_name": "com.scottyab.rootbeer.sample.CheckRootTask$OnCheckRootFinishedListener",
+  "_methods": [
+    "constructor",
+    "class",
+    "onCheckRootFinished"
+  ],
+  "_fields": []
+}
+...
+
+```
+
+Given the verbosity of the output, the system classes can be filtered out programmatically to make output more readable and relevant to the use case.
+
+###### Getting Loaded Libraries
+
+In Frida REPL process related information can be obtained using the `Process` command. Within the `Process` command the function `enumerateModules` lists the libraries loaded into the process memory.
+
+```
+[Huawei Nexus 6P::sg.vantagepoint.helloworldjni]-> Process.enumerateModules()
+[
+    {
+        "base": "0x558a442000",
+        "name": "app_process64",
+        "path": "/system/bin/app_process64",
+        "size": 32768
+    },
+    {
+        "base": "0x78bc984000",
+        "name": "libandroid_runtime.so",
+        "path": "/system/lib64/libandroid_runtime.so",
+        "size": 2011136
+    },
+    {
```

Comment:
Let's keep the output always to a minimum, so it's not overwhelming and the reader can focus on what's important.  Can you please just list the first two loaded libraries?

Reply:
I am sorry, seems i always leave more than what is needed.